advertise/addr.go

```diff
@@ -5,99 +5,164 @@ package advertise
 
-// FirstAddress returns the first IPv4 address from the given interface names.
-// Addresses used for APIPA will be ignored if possible.
-func FirstAddress(interfaces []string) (net.IP, error) {
+const AllInterfaces = "all"
+
+// FirstAddress returns the first IPv4/IPv6 address from the given interface names.
+// Link-local unicast addresses will be ignored if possible.
+func FirstAddress(interfaces []string) (string, error) {
+	return firstAddress(interfaces, getInterfaceAddresses, net.Interfaces)
+}
+
+// NetworkInterfaceAddressGetter matches the signature of net.InterfaceByName() to allow for test mocks.
+type NetworkInterfaceAddressGetter func(name string) ([]netip.Addr, error)
+
+// InterfaceLister matches the signature of net.Interfaces() to allow for test mocks.
+type InterfaceLister func() ([]net.Interface, error)
+
+// FirstAddress returns the first IPv4/IPv6 address from the given interface names.
+// Link-local unicast addresses will be ignored if possible.
+func firstAddress(interfaces []string, interfaceAddrsFunc NetworkInterfaceAddressGetter, interfaceLister InterfaceLister) (string, error) {
 	var (
-		errs      *multierror.Error
-		privateIP net.IP
+		errs   *multierror.Error
+		bestIP netip.Addr
 	)
 
+	if len(interfaces) == 1 && interfaces[0] == AllInterfaces {
+		infs, err := interfaceLister()
+		if err != nil {
+			return "", fmt.Errorf("failed to get interface list: %w", err)
+		}
+		interfaces = make([]string, len(infs))
+		for i, v := range infs {
+			interfaces[i] = v.Name
+		}
+	}
+
 	for _, ifaceName := range interfaces {
-		iface, err := net.InterfaceByName(ifaceName)
+		addrs, err := interfaceAddrsFunc(ifaceName)
 		if err != nil {
 			err = fmt.Errorf("interface %q: %w", ifaceName, err)
 			errs = multierror.Append(errs, err)
 			continue
 		}
 
-		addrs, err := iface.Addrs()
-		if err != nil {
-			err = fmt.Errorf("interface %q addrs: %w", ifaceName, err)
-			errs = multierror.Append(errs, err)
-			continue
-		} else if len(addrs) <= 0 {
-			err = fmt.Errorf("interface %q has no addresses", ifaceName)
-			errs = multierror.Append(errs, err)
+		canditate := filterBestIP(addrs)
+		if !canditate.IsValid() {
 			continue
 		}
 
-		foundAddr := findSuitableIP(addrs)
-		if foundAddr == nil {
-			err = fmt.Errorf("interface %q has no suitable addresses", ifaceName)
-			errs = multierror.Append(errs, err)
-			continue
+		if canditate.Is4() && !canditate.IsLinkLocalUnicast() {
+			// Best address possible, we can return early.
+			return canditate.String(), nil
 		}
 
-		if !IsAutomaticPrivateIP(foundAddr) {
-			return foundAddr, nil
-		} else if privateIP == nil {
-			privateIP = foundAddr
+		bestIP = filterBestIP([]netip.Addr{canditate, bestIP})
+	}
+	if !bestIP.IsValid() {
+		if errs != nil {
+			return "", errors.Wrapf(errs, "no useable address found for interfaces %v", interfaces)
+		} else {
+			return "", fmt.Errorf("no useable address found for interfaces %v", interfaces)
```

Comment:
Let's not add a new dependency but [wrap the error](https://go.dev/blog/go1.13-errors#:~:text=Wrapping%20errors%20with%20%25w) with `%w` instead

```suggestion
	if !bestIP.IsValid() {
		if errs.ErrorOrNil() != nil {
			return "", fmt.Errorf(errs, "no useable address found for interfaces %v: %w", interfaces, errs.ErrorOrNil())
		} else {
			return "", fmt.Errorf("no useable address found for interfaces %v", interfaces)
```

I'm on the fence whether the branch is needed here, but let's have it for now.